---
title: "Untitled"
format: pdf
pdf-engine: lualatex
editor: visual
mainfont: Calibri
monofont: PT Mono
always_allow_html: yes
header-includes:
   \usepackage[dvipsnames]{xcolor}
   \definecolor{darkblue}{rgb}{0.0, 0.0, 0.55}
   \definecolor{ivory}{rgb}{1.0, 1.0, 0.94}
---

### Preprocessing reads

##### Preparing reads

Dorado doesn't support demultiplexing of dual indexes on 5' and 3' ends. Additionally, ligated library reads can be either orientation. Most straightforward approach is to solve this, is to utilise flanking adapter sequences and cutadapt.

Extracting forward reads to fastq file can be performed with following command

*cutadapt -g t2f...rc(t2r) -O 13 --trimmed-only -m 1300 -M 1650 -o forward_out.fastq.gz raw_reads.fastq.gz*

Extracting reverse reads

*cutadapt -g t2r...rc(t2f) -O 13 -m 1300 -M 1650 --trimmed-only -o reverse_out.fastq.gz raw_reads.fastq.gz*

**Note!** e=0, minimum (m) and maximum size limits reduce chance of misaligned reads

Next step is to reverse complement reverse read file and join all reads to single file. This will make life easier later in the data analysis workflow.

*seqkit seq -rp --seq-type DNA reverse_out.fastq.gz \| gzip \>reverse_comp.fastq.gz*

Next step is to merge two files together

*cat forward_out.fastq.gz reverse_comp.fastq.gz \>nanopore_reads.fastq.gz*

##### Demultiplexing reads

Prepare list of barcodes as a fasta file

Use cutadapt to demux fastq file. In example output files are written to demuxed subdirectory

*cutadapt -e 0 -O 12 -g file:barcodes.fasta -o "demuxed/{name}.fastq.gz" input.fastq.gz*

##### Trimming PCR amplification primers

You can use cutadapt and bash scripts to trim forward and revers PCR primers from demultiplexed sequence files
